added verification check for misspelled propTypes

M-ZubairAhmed opened the pull request:
### Prerequisites 
- [X] All test ran successfully
- [X] Code formatted with prettier
- [X] Code linting done
- [X] Flow run completed
- [X] completed the CLA.

## Changes
Added the check if `PropTypes` is entered instead of `propTypes` in the static method of class for props types check.
*Please let me know if i have to change anything*

M-ZubairAhmed commented:
solves https://github.com/facebook/react/issues/11510

gaearon commented on packages/react/src/ReactElementValidator.js:
```diff
@@ -221,7 +221,14 @@ function validatePropTypes(element) {
   var name = componentClass.displayName || componentClass.name;
   var propTypes = componentClass.propTypes;
-
+  if (componentClass.PropTypes !== undefined) {
```
We should deduplicate this warning so it doesn't warn more than once. See how we do with most other warnings.

M-ZubairAhmed replied:
I have added the boolean flag after observing other places where `warning()` was used. Also tested locally

M-ZubairAhmed commented:
I hope this is the correct approach if not I will gladly change it.

gaearon commented on packages/react/src/ReactElementValidator.js:
```diff
@@ -214,14 +214,24 @@ function validateChildKeys(node, parentType) {
   var name = componentClass.displayName || componentClass.name;
   var propTypes = componentClass.propTypes;
-
+  if (
```
Maybe let's put it into an `else` clause to the next `if`?

```js
if (propTypes) {
  // ...
} else if (componentClass.PropTypes && !propTypesMisspellWarningShown) {
  // ...
}
```

M-ZubairAhmed replied:
I have also tested this `fixtures/dom` by introducing some props and misspelling `propTypes`

gaearon commented on packages/react/src/ReactElementValidator.js:
```diff
@@ -214,14 +214,24 @@ function validateChildKeys(node, parentType) {
+    propTypesMisspellWarningShown = true;
+    warning(
+      false,
+      'Static propTypes method should be accessed by `.propTypes = { }` instead of `.PropTypes = { }`',
```
Let's reword to say:

>Component %s declared `PropTypes` instead of `propTypes`. Did you misspell the property assignment?

And pass the component name `name  || 'Unknown'` as an argument.

M-ZubairAhmed replied:
changes done. Thank you @gaearon for being so patient with me.

gaearon commented on packages/react/src/ReactElementValidator.js:
```diff
@@ -214,18 +214,28 @@ function validateChildKeys(node, parentType) {
  *
  * @param {ReactElement} element
  */
+var propTypesMisspellWarningShown;
```
Can we move this up to other DEV only variables?

gaearon replied:
Also let's initialise it to false.

M-ZubairAhmed replied:
Added the variable declaration under `if(__DEV__){` which is after import statements. Now working on test.

M-ZubairAhmed replied:
Should the test be added in both `ReactElementValidator-test` and `ReactJSXElementValidator-test`?

gaearon replied:
I think just one is fine.

M-ZubairAhmed replied:
Oh i just added test in both of them, i pushed my code. Should i remove from one ?
Also i have verified all tests were passing and ran build in my local example which was reproducing the error.
I was wondering if i should add another test it(`should not show any errors is correct propTypes property assignment is padded)`?

M-ZubairAhmed commented:
are the tests correct or should I change something?

M-ZubairAhmed commented:
friendly ping!

gaearon commented on packages/react/src/__tests__/ReactElementValidator-test.js:
```diff
@@ -434,6 +434,29 @@ describe('ReactElementValidator', () => {
+    );
+
+    expectDev(console.error.calls.count()).toBe(1);
+    /*eslint-disable max-len */
```
Nit: instead of disabling, you can split the message into multiple lines and concatenate them. Other tests do this.

M-ZubairAhmed replied:
ok i removed eslint disable and added the split in string.

M-ZubairAhmed commented:
ok i removed eslint disable and added the split in string.

M-ZubairAhmed commented:
as per recent changes in master i have added __DEV__ check to my tests. I hope it is correct?

gaearon commented:
This mostly looks good. I think we'll want to try this internally and see if it fires any unexpected warnings at FB. (e.g. I think I saw `.PropTypes` used intentionally on something renderable in RN. But maybe that was in the past.)

M-ZubairAhmed commented:
thanks do you want me to run a quick check on a react Native project?

gaearon commented:
Would be nice.

M-ZubairAhmed commented:
Hey @gaearon i had a chance to run the latest master build in a rn project:
*There was no un usual error produced*

##### Steps
- Build latest from master branch
- replaced the build version of react inside RN node modules
- Added `prop-types` to RN project
- kept track of console logs to check if anything showed up
- intentionally passed `PropTypes` instead of `propTypes`
- correct warning was shown as expected
- reverted the propTypes declaration
- no error was shown

Does this test is good?
